<a href="https://colab.research.google.com/github/jawwadturk/CodSoft-Internship-Task1/blob/main/AI_Assignment01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

# Load the dataset

In [2]:
data = fetch_openml(name='spambase', version=1)

/usr/local/lib/python3.10/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


# Split data into features (X) and labels (y)

In [3]:
X, y = data.data, data.target.astype(int)

# Step 2: Split the Data

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 3: Apply Various Classification Methods
# Method 1: Logistic Regression

In [5]:
lr_model = LogisticRegression(max_iter=1000)
lr_model.fit(X_train, y_train)
lr_train_accuracy = accuracy_score(y_train, lr_model.predict(X_train))
lr_test_accuracy = accuracy_score(y_test, lr_model.predict(X_test))

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# Method 2: Support Vector Machine (SVM)

In [6]:
svm_model = SVC()
svm_model.fit(X_train, y_train)
svm_train_accuracy = accuracy_score(y_train, svm_model.predict(X_train))
svm_test_accuracy = accuracy_score(y_test, svm_model.predict(X_test))

# Method 3: Random Forest

In [7]:
rf_model = RandomForestClassifier()
rf_model.fit(X_train, y_train)
rf_train_accuracy = accuracy_score(y_train, rf_model.predict(X_train))
rf_test_accuracy = accuracy_score(y_test, rf_model.predict(X_test))

# Method 4: Naive Bayes

In [8]:
nb_model = GaussianNB()
nb_model.fit(X_train, y_train)
nb_train_accuracy = accuracy_score(y_train, nb_model.predict(X_train))
nb_test_accuracy = accuracy_score(y_test, nb_model.predict(X_test))

# Step 4: Compare Results

In [9]:
results = {
    'Logistic Regression': (lr_train_accuracy, lr_test_accuracy),
    'SVM': (svm_train_accuracy, svm_test_accuracy),
    'Random Forest': (rf_train_accuracy, rf_test_accuracy),
    'Naive Bayes': (nb_train_accuracy, nb_test_accuracy)
}

# Print results

In [10]:
for method, (train_acc, test_acc) in results.items():
    print(f"{method} - Training Accuracy: {train_acc:.4f}, Testing Accuracy: {test_acc:.4f}")

Logistic Regression - Training Accuracy: 0.9307, Testing Accuracy: 0.9197
SVM - Training Accuracy: 0.7136, Testing Accuracy: 0.6623
Random Forest - Training Accuracy: 0.9995, Testing Accuracy: 0.9566
Naive Bayes - Training Accuracy: 0.8220, Testing Accuracy: 0.8208


**Based on the provided results, the Random Forest method has the highest testing accuracy (0.9522), followed by Logistic Regression (0.9197), Naive Bayes (0.8208), and SVM (0.6623).**